### ISSUE TITLE
TexutreRect renders as black when Stretch Mode is set to Tile on GLES2

### ISSUE THREAD

**Issue by Tekuzo** [godotengine/godot#47955]
**Godot version:**

3.2.3 and 3.3 RC9

**OS/device including version:**

iPhone SE (2nd version) running iOS 14.4.2. This bug was discovered using GLES2. When I switched rendering to GLES3 I did not encounter the bug.

**Issue description:**

A TextureRect when put into tile mode will only render as black. And not simply the texture, but the whole area of the texture rect will render as black.
![IMG_0151](https://user-images.githubusercontent.com/6944363/115037020-0885b800-9e9c-11eb-9a8a-4b65e8abc756.PNG)

**Steps to reproduce:**
Make a scene with a texture rect, switch its stretch mode to tile.

**Minimal reproduction project:**

I have created a sample project and removed my Team ID and Bundle ID (I hope)
[Sample Project.zip](https://github.com/godotengine/godot/files/6325642/Sample.Project.zip)

**Comment by Calinou** [comment]
See https://github.com/godotengine/godot/issues/40373. This is supposed to be fixed, but apparently it's not working?

Can you reproduce this issue with GLES2 on your own desktop machine?

**Comment by Tekuzo** [comment]
No, I can not re-produce this on Windows, Mac, Linux or Android, it only seems to happen on GLES2 on iOS

**Comment by volzhs** [comment]
maybe similar issue with #32837?

**Comment by Tekuzo** [comment]
If the stretch mode is on Scale then the issue goes away as well.

**Comment by lawnjelly** [comment]
> maybe similar issue with #32837?

If so, that bug was fixed by setting `rendering/gles2/compatibility/disable_half_float`. I haven't yet looked at the shader, but in the meantime this setting is worth a try.

**Comment by Tekuzo** [comment]
> > maybe similar issue with #32837?
> 
> If so, that bug was fixed by setting `rendering/gles2/compatibility/disable_half_float`. I haven't yet looked at the shader, but in the meantime this setting is worth a try.

I tried to disable this feature and there was no change I have attached a screenshot to demonstrate.

![IMG_0154](https://user-images.githubusercontent.com/6944363/115064001-5c06fe80-9eba-11eb-8f6b-fa067e371ba8.PNG)

**Comment by lawnjelly** [comment]
Does it work with a power of 2 texture size?

**Comment by Tekuzo** [comment]
> Does it work with a power of 2 texture size?

Yes
I made the texture 16 by 16 pixels and it rendered

**Comment by lawnjelly** [comment]
Ok that confirms the general area to look. I'll try and take a look tomorrow.

Incidentally, I would always recommend using POT textures on such hardware regardless, because it will usually run faster, and tiling in the fragment shader can be very slow / problematic on some hardware.

**Comment by lawnjelly** [comment]
When I force `support_npot_repeat_mipmap` to false, and emulate what might be happening on npot hardware on desktop, it does work correctly for me (that was #40373) so this is a different bug.

You could try it with batching disabled, I suspect it will not work either.

This may be rather difficult to solve without the hardware in question. It may be something in the force repeat shader it doesn't like.

**Comment by Tekuzo** [comment]
Should I try this with batching disabled on 3.3 stable now?

**Comment by lawnjelly** [comment]
> Should I try this with batching disabled on 3.3 stable now?

Yup sure. That will enable us to know better the rough area that is failing.

**Comment by Tekuzo** [comment]
I just tried with batching disabled on 3.3 stable

exported an iOS project, ran on my iPhone

No change. still rendered as a black box.
